### ISSUE TITLE
packer 0.7.5 - Docker builder - shell provisioner cannot start the script

### ISSUE THREAD

**Issue by pwong-mapr** [mitchellh/packer#2040]
Host machine info: 
* `lsb_release -a = CentOS release 6.6 (Final)`
* `docker --version = Docker version 1.5.0, build a8a31ef/1.5.0`
* `packer version = Packer v0.7.5`

I am using this exact file: https://gist.github.com/pwong-mapr/774d7b9e0501ba26bcf1

Packer hangs at "Provisioning with shell script" for a long time, then fails. The complete log is here: https://gist.github.com/pwong-mapr/3c487edc67d2fdb03312

Seems like the problem is that it either cannot mount the script correctly in the container, or it cannot run the script, or it may not even be able to send inputs.

Cannot reproduce the issue with pure-Docker methods.

Doing `docker attach <container ID>` while packer is in the middle of the build, then doing `exit` to stop the container, does not make packer exit early.

Doing `docker attach <container ID>` while packer is in the middle of the build, then doing `ls /packer-files` shows nothing resembling the script.

**Comment by thrashr888** [comment]
I have the same issue with boot2docker on OSX with the same versions of Docker and Packer. I get the same output.

**Comment by thrashr888** [comment]
After a little debugging with `PACKER_LOG=1` on, `docker run -v /Users/thrashr888/tmp/packer-docker393691987:/packer-files -d -i -t ubuntu:latest /bin/bash` is the command that's run. In my case, the local folder `/Users/thrashr888/tmp/packer-docker393691987` is empty.

**Comment by thrashr888** [comment]
`ls /packer-files/` actually contains the cmd and upload files but they're removed then replaced during this loop:

```
2015/04/17 16:51:22 packer-provisioner-shell: 2015/04/17 16:51:22 [INFO] 46 bytes written for 'uploadData'
2015/04/17 16:51:22 [INFO] 46 bytes written for 'uploadData'
2015/04/17 16:51:22 packer-builder-docker: 2015/04/17 16:51:22 Executing in container 7d1261a8401cf953883525473460e65b0db858fd9aff771451269dfc09a73236: "(cp /packer-files/upload007033354 /tmp/script.sh) >/packer-files/cmd126622177 2>&1; echo $? >/packer-files/cmd126622177-exit"
2015/04/17 16:51:24 packer-provisioner-shell: 2015/04/17 16:51:24 Retryable error: Error uploading script: Upload failed with non-zero exit status: 1
```

**Comment by ngrennan-inflection** [comment]
I am the same issue.

packer version:
built from git master as of Apr 20

docker version:
1.6.0

Host distribution:
Fedora 21

Container distribution:
CentOS 7

**Comment by thrashr888** [comment]
i'm using the patched version from #1752 and it works just fine.

https://www.dropbox.com/s/4v5jvvxj1k5mpst/packer-osx-patched-0.7.5.zip?dl=1

**Comment by thrashr888** [comment]
Similar to reports on #1752, I can confirm that the patch doesn't work for remote Docker servers. It only works locally.

**Comment by jamescarr** [comment]
I have the same problem:

Packer 0.7.5
Docker 1.6.0
Ubuntu 14.04

**Comment by maguec** [comment]
I'm noticing that the files seem to get created and removed several times during retries:

```
drwxr-xr-x 33 root root 4096 May 12 19:58 ..
-rw-------  1 1000 1000  117 May 12 20:03 upload056472909
-rw-------  1 1000 1000    0 May 12 20:03 cmd763017800
drwx------  2 1000 1000 4096 May 12 20:03 .
root@041d530d39ec:/packer-files# ls -altr
total 8
drwxr-xr-x 33 root root 4096 May 12 19:58 ..
drwx------  2 1000 1000 4096 May 12 20:03 .
root@041d530d39ec:/packer-files# ls -altr
total 12
drwxr-xr-x 33 root root 4096 May 12 19:58 ..
-rw-------  1 1000 1000  117 May 12 20:03 upload941850887
-rw-------  1 1000 1000    0 May 12 20:03 cmd492638394
drwx------  2 1000 1000 4096 May 12 20:03 .
root@041d530d39ec:/packer-files# ls -altr
total 8
drwxr-xr-x 33 root root 4096 May 12 19:58 ..
drwx------  2 1000 1000 4096 May 12 20:03 .
```

Packer 0.7.5
Docker 1.6.1
Ubuntu 15.04

```
{
  "provisioners": [
    {
      "type"   : "shell",
      "inline" : [
        "sudo apt-get update && sudo apt-get -y upgrade",
        "sudo apt-get -y install puppet",
        "sudo mkdir -p /tmp/hieradata"
      ]
    }
  ],
  "builders": [
    {
      "type"        : "docker",
      "image"       : "ubuntu",
      "export_path" : "img.tar"
    }
  ],
  "post-processors": [
  ]
}
```

**Comment by deimosfr** [comment]
Hi,

Got the same issue:
Docker 1.6.1
Debian Jessie
Packer 0.7.5

**Comment by ahmadnazir** [comment]
Reproduced this error for Docker version ```1.4.1``` and ```1.6.2``` when trying to provision a docker container using the puppet provisioner.

However, **the problem does not occur for version ```1.3.3```** .

Host machine info:

```
$ lsb_release -a
..
Distributor ID:	Ubuntu
Description:	Ubuntu 14.04.2 LTS
Release:	14.04
Codename:	trusty
```

Output with ```PACKER_LOG=1```

```
..
==> docker: Provisioning with Puppet...
2015/05/25 23:27:10 ui:     docker: Creating Puppet staging directory...
    docker: Creating Puppet staging directory...
2015/05/25 23:27:10 packer-builder-docker: 2015/05/25 23:27:10 Executing in container b877d3235df13fd8e51f2303dd491367e96a3d8f2835f9994b7818c02326299a: "(mkdir -p '/tmp/packer-puppet-masterless') >/packer-files/cmd617401699 2>&1; echo $? >/packer-files/cmd617401699-exit"
2015/05/25 23:27:12 packer-builder-docker: 2015/05/25 23:27:12 [INFO] RPC endpoint: Communicator ended with: 1
2015/05/25 23:27:12 [INFO] 0 bytes written for 'stderr'
2015/05/25 23:27:12 [INFO] 0 bytes written for 'stdout'
2015/05/25 23:27:12 [INFO] RPC client: Communicator ended with: 1
2015/05/25 23:27:12 [INFO] RPC endpoint: Communicator ended with: 1
2015/05/25 23:27:12 packer-provisioner-puppet-masterless: 2015/05/25 23:27:12 [INFO] 0 bytes written for 'stderr'
2015/05/25 23:27:12 packer-provisioner-puppet-masterless: 2015/05/25 23:27:12 [INFO] 0 bytes written for 'stdout'
2015/05/25 23:27:12 packer-provisioner-puppet-masterless: 2015/05/25 23:27:12 [INFO] RPC client: Communicator ended with: 1
2015/05/25 23:27:12 ui: ==> docker: Killing the container: b877d3235df13fd8e51f2303dd491367e96a3d8f2835f9994b7818c02326299a
==> docker: Killing the container: b877d3235df13fd8e51f2303dd491367e96a3d8f2835f9994b7818c02326299a
2015/05/25 23:27:12 ui error: Build 'docker' errored: Error creating staging directory: Non-zero exit status.
2015/05/25 23:27:12 Builds completed. Waiting on interrupt barrier...
2015/05/25 23:27:12 machine readable: error-count []string{"1"}
2015/05/25 23:27:12 ui error: 
==> Some builds didn't complete successfully and had errors:
2015/05/25 23:27:12 machine readable: docker,error []string{"Error creating staging directory: Non-zero exit status."}
2015/05/25 23:27:12 ui error: --> docker: Error creating staging directory: Non-zero exit status.
2015/05/25 23:27:12 ui: 
==> Builds finished but no artifacts were created.
2015/05/25 23:27:12 waiting for all plugin processes to complete...
Build 'docker' errored: Error creating staging directory: Non-zero exit status.
```

**Comment by mitchellh** [comment]
Fixed with #1993
